src/coreclr/jit/emit.cpp

```diff
@@ -8208,76 +8208,92 @@ CORINFO_FIELD_HANDLE emitter::emitSimdConst(simd_t* constValue, emitAttr attr)
 //
 void emitter::emitSimdConstCompressedLoad(simd_t* constValue, emitAttr attr, regNumber targetReg)
 {
-    assert(EA_SIZE(attr) >= 8);
-    unsigned    cnsSize = EA_SIZE(attr);
-    instruction ins     = (cnsSize == 8) ? INS_movsd_simd : INS_movups;
+    assert(EA_SIZE(attr) >= 8 && EA_SIZE(attr) <= 64);
 
-    for (unsigned size = 4; size < cnsSize; size *= 2)
+    unsigned    cnsSize  = EA_SIZE(attr);
+    unsigned    dataSize = cnsSize;
+    instruction ins      = (cnsSize == 8) ? INS_movsd_simd : INS_movups;
+
+    // Most constant vectors tend to have repeated values, so we will first check to see if
+    // we can replace a full vector load with a smaller broadcast.
+
+    if ((dataSize == 64) && (constValue->v256[1] == constValue->v256[0]))
     {
-        // If the constant has zero upper elements/lanes, we can use a smaller load,
-        // because all scalar and vector loads from memory zero the upper.
+        assert(emitComp->IsBaselineVector512IsaSupportedDebugOnly());
+        dataSize = 32;
+        ins      = INS_vbroadcastf32x8;
+    }
 
-        simd_t val = {};
-        memcpy(&val, constValue, size);
+    if ((dataSize == 32) && (constValue->v128[1] == constValue->v128[0]))
+    {
+        assert(emitComp->IsBaselineVector256IsaSupportedDebugOnly());
+        dataSize = 16;
+        ins      = INS_vbroadcastf128;
+    }
 
-        if (memcmp(&val, constValue, cnsSize) == 0)
+    if ((dataSize == 16) && (constValue->u64[1] == constValue->u64[0]))
+    {
+        if (((cnsSize == 16) && emitComp->compOpportunisticallyDependsOn(InstructionSet_SSE3)) ||
+            emitComp->compOpportunisticallyDependsOn(InstructionSet_AVX))
         {
-            switch (size)
-            {
-                case 4:
-                    ins = INS_movss;
-                    break;
-                case 8:
-                    ins = INS_movsd_simd;
-                    break;
-                default:
-                    ins = INS_movups;
-                    break;
-            }
-
-            cnsSize = size;
-            attr    = EA_ATTR(size);
-            break;
+            dataSize = 8;
+            ins      = (cnsSize == 16) ? INS_movddup : INS_vbroadcastsd;
         }
+    }
 
-        if (((size == 8) && (cnsSize == 16) && emitComp->compOpportunisticallyDependsOn(InstructionSet_SSE3)) ||
-            emitComp->compOpportunisticallyDependsOn(InstructionSet_AVX))
+    if ((dataSize == 8) && (cnsSize >= 16) && (constValue->u32[1] == constValue->u32[0]))
```

Comment:
This is a bit easy to miss and so a small comment explaining that we can encounter `cnsSize == 8` for `TYP_DOUBLE` or `TYP_SIMD8` (i.e. Vector2) and in such a case we can't broadcast since SIMD registers are 16/32/64 in size would be helpful for future readers who aren't as familiar with the specifics.

Reply:
If it's ok, I'll tack a comment on in my next PR unless I need to churn CI again for another reason.